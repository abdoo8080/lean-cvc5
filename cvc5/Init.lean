/-
Copyright (c) 2023-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed, Adrien Champion
-/

import Lean.Elab.Command

namespace cvc5

/-! ## DSL for definition  DRY -/
section defsMacro

open Lean
open Elab
open Command (CommandElab CommandElabM)

declare_syntax_cat externKw

declare_syntax_cat defsItem
declare_syntax_cat defsItemHead
declare_syntax_cat defsItemTail!?
declare_syntax_cat defsItemTail

declare_syntax_cat defsMod

/-- Arity of an expression.

Stolen from [batteries].

[batteries]: https://leanprover-community.github.io/mathlib4_docs/Batteries/Lean/Expr.html#Lean.Expr.forallArity
-/
def forallArity : Expr → Nat
  | .mdata _ b => forallArity b
  | .forallE _ _ body _ => 1 + forallArity body
  | _ => 0

syntax "?" : defsMod
syntax "!" : defsMod
syntax "!?" : defsMod
syntax "?!" : defsMod

scoped syntax (name := defsItemStxHead)
  declModifiers
  ("@[" "force" str "]")?
  "def " defsMod ? ident
: defsItemHead

scoped syntax (name := defsItemStxTail!?)
  withPosition(ppLine "with!? "
    group(
      colGt
      docComment ?
      ident
    )*
  )
: defsItemTail!?

scoped syntax (name := defsItemStxTail)
  withPosition(ppLine "with "
    group(
      colGt
      docComment ?
      declId optDeclSig ":= " withPosition(group(colGe term))
    )*
  )
: defsItemTail

scoped syntax (name := defsItemStx)
  defsItemHead
  declSig
  (defsItemTail!?)?
  (defsItemTail)?
: defsItem

unsafe def elabDefsItem
  (pref : String) (forceMods : Option (TSyntax `defsMod))
: CommandElab
| `(defsItem|
    $mods:declModifiers
    $[ @[ force $forcedName ] ]?
    def $ident:ident $identSig:declSig
    $[ with!? $[
        $[$autoDoc]?
        $autoId
    ]* ]?
    $[ with $[
        $[$subDoc]?
        $subId $subSig := $subDef
    ]* ]?
) => do
  if let some defMods := forceMods then
    let stx ← `(defsItem|
$mods:declModifiers
$[ @[ force $forcedName ] ]?
def $defMods $ident:ident $identSig:declSig
$[ with!? $[
    $[$autoDoc]?
    $autoId
]* ]?
$[ with $[
    $[$subDoc]?
    $subId $subSig := $subDef
]* ]?
    )
    return ← elabDefsItem pref none stx

  let externName :=
    let id :=
      if let some forcedName := forcedName then
        forcedName.getString
      else
        ident.getId.toString
    pref ++ "_" ++ id |> Syntax.mkStrLit
  let mods ←
    match mods with
    | `(Parser.Command.declModifiersT|
      $[$doc:docComment]? $[@[ $[ $attrs ],* ]]? $[$vis]? $[$isNc]? $[$isUnsafe]? $[$opt]?
    ) => do
      let ext ← `(Parser.Term.attrInstance| extern $externName:str)
      let attrs := attrs.getD #[] |>.push ext
      `(Parser.Command.declModifiersT|
        $[$doc:docComment]? @[ $[$attrs],* ] $[$vis]? $[$isNc]? $[$isUnsafe]? $[$opt]?
      )
    | _ => throwUnsupportedSyntax
  let mainDef ←`(
    set_option linter.unusedVariables false in
    $(⟨mods⟩):declModifiers
    opaque $ident:declId $identSig
  )
  Command.elabCommand mainDef

  let fullName :=
    Lean.Name.mkSimple ident.getId.toString
    |> (← Elab.liftMacroM Macro.getCurrNamespace).append
  let fullIdent := Lean.mkIdent fullName

  let define doc? id sig? (body : Syntax.Term) : CommandElabM _ := do
    if let some doc := doc? then
      `(command|
        $doc:docComment
        def $id:declId $sig?:optDeclSig := $body
      )
    else
      `(command|
        @[inherit_doc $fullIdent]
        def $id:declId $sig?:optDeclSig := $body
      )

  if let (some autoDoc?, some autoId) := (autoDoc, autoId) then
    let env ← getEnv
    let arity ←
      if let some (.opaqueInfo i) := env.find? fullName then
        pure (forallArity i.type)
      else
        throwError s!"failed to retrieve arity of (opaque) function `{ident}`"

    let mut args := Array.empty
    for i in [0:arity] do
      let arg := Lean.Name.mkSimple s!"v{i}" |> Lean.mkIdent
      args := args.push arg
    let funCall : TSyntax `term ← `(term| ( $fullIdent $[ $args ]* ))

    for (autoDoc?, autoId) in autoDoc?.zip autoId do
      let id : String := autoId.getId.toString
      let body ←
        if id.endsWith "!" then
          let unwrapName := Lean.Name.mkStr3 "cvc5" "Error" "unwrap!" |> Lean.mkIdent
          `(fun $[$args]* => $funCall |> $unwrapName)
        else if id.endsWith "?" then
          `(fun $[$args]* => $funCall |> Except.toOption)
        else
          throwError s!"unexpected auto function name `{id}`: expected `<ident>!` or `<ident>?`"
      let cmd ← define autoDoc? autoId (← `(optDeclSig|)) body
      Command.elabCommand cmd

  if let
    (some subDoc?, some subId, some subSig, some subDef)
    := (subDoc, subId, subSig, subDef)
  then
    let all := subDoc?.zip subId |>.zip subSig |>.zip subDef
    for (((subDoc?, subId), subSig), subDef) in all do
      Command.elabCommand
        (← define subDoc? subId subSig subDef)
| `(defsItem|
    $mods:declModifiers
    $[ @[ force $forcedName ] ]?
    def $defsMod $ident:ident $identSig:declSig
    $[ with $[
        $[$subDoc]?
        $subId $subSig := $subDef
    ]* ]?
) => do
  let name := ident.getId
  let identOpt _ := name.appendAfter "?" |> Lean.mkIdent
  let identPanic _ := name.appendAfter "!" |> Lean.mkIdent
  let auto : Array Ident ←
    match defsMod with
    | `(defsMod| ?) => pure #[identOpt ()]
    | `(defsMod| !) => pure #[identPanic ()]
    | `(defsMod| ?!)
    | `(defsMod| !?) => pure #[identOpt (), identPanic ()]
    | _ => throwUnsupportedSyntax
  let stx ← `(defsItemStx|
    $mods:declModifiers
    $[ @[ force $forcedName ] ]?
    def $ident:ident $identSig:declSig
    with!? $[ $auto:ident ]*
    $[ with $[
        $[ $subDoc:docComment ]?
        $subId:declId $subSig:optDeclSig := $subDef:term
    ]* ]?
  )
  elabDefsItem pref none stx
| _ => throwUnsupportedSyntax

/-- Defines similar functions realized by `extern`.

```
extern! in "prefix"
  /-- Create a Boolean constant.

  - `b`: The Boolean constant.

  Will create an opaque definition with `[@extern extStr]` where
  `extStr = "prefix" ++ "_" ++ "myFunction"`.
  -/
  def myFunction : Term → Except Error Op
  with!?
    endsWithBang!
    endWithQuestion?
  with
    myOtherFunction : Term → Op :=
      Error.unwrap! ∘ myFunction
    /-- Optional function docstring: if none, inherit from the main function. -/
    yetAnotherFunction : Term → Option Op :=
      Except.toOption ∘ myFunction
```

- `in "prefix"` is optional; if none, then the prefix will be the (last component of the) name of
  the current namespace with the first letter lowercased. Fails if the current namespace has no
  components.

- `with ...`: takes a sequence of identifiers, each generate a function that
  - unwraps the result if `!`-ended, which generates code similar to `myOtherFunction` above;
  - turns a result into an option if `?`-ended, which generates code similar to `yetAnotherFunction`
    above;
  - fails otherwise.

  The `with ...` syntax is currently only compatible with external functions that produce `Except
  Error α` values.

- Supports `declModifiers` on the main (`def`) function `myFunction` such as `private`.
- Accepts a list of external (`def`) functions, each with its `with` clauses.

This macro can generate optional (`?`) and panic (`!`) wrappers even more automatically.

```
extern! "prefix"
  /-- Create a Boolean constant.

  - `b`: The Boolean constant.

  Will create an opaque definition with `[@extern extStr]` where
  `extStr = "prefix" ++ "_" ++ "myFunction"`.
  -/
  def !? myFunction : Term → Except Error Op
  with
    myOtherFunction : Term → Op :=
      myFunction!
    /-- Optional function docstring: if none, inherit from the main function. -/
    yetAnotherFunction : Term → Option Op :=
      myFunction?
```

Notice the `!?` between `def` and `myFunction`. This generates `myFunction!` and `myFunction?` which
respectively panic-unwrap errors and turn the `Except` into an `Option`. In other words, it is the
same as (and internally turned into) a `with!? myFunction! myFunction?` clause using the first
syntax above.

Besides `!?`, the following are also supported:
- `?!`: same behavior as `!?`;
- `!`: only generate the panic unwrapper;
- `?`: only generate the `Option` unwrapper.
-/
scoped syntax (name := multidefs)
  withPosition("external! " ("in " str)? ppLine group(colGt defsItem)+)
: command

@[inherit_doc multidefs, command_elab multidefs]
unsafe def multidefsImpl : CommandElab
| `(command|
  external! $[in $pref:str]? $[$defsItems]*
) => do
  let ns ← getCurrNamespace
  let pref ←
    if let some pref := pref then
      pure pref.getString
    else
      -- println! "componentsRev for {ns.toString}"
      -- for c in ns.componentsRev do
      --   println! "- {c.toString}"
      if let super::_ := ns.componentsRev then
        let mut super := super.toString
        if 0 < super.length then
          super := super.get 0 |>.toLower |> super.set 0
        -- println! "-> `{super}`"
        pure super
      else
        throwError "failed to retrieve current workspace, please provide an explicit prefix"
  for defsItem in defsItems do
    elabDefsItem pref none defsItem
| _ => throwUnsupportedSyntax

scoped syntax (name := externkw)
  ("extern_def " <|> "extern_def! " <|> "extern_def? " <|> "extern_def!? " <|> "extern_def?! ")
: externKw

/-- Defines an external, opaque function with optional helpers.

```
/-- Some documentation. -/
extern_def!? in "prefix" myFunction : Term → Except Error Op
```

Generates an opaque definition

```
/-- Some documentation. -/
@[extern "prefix_myFunction"]
def myFunction : Term → Except Error Op
```

- `in "prefix"` is optional, uses the current namespace with first letter lowercased if none.

- `extern_def!?` also defines `myFunction! : Term → Op` and `myFunction? : Term → Option Op` in
  terms of `myFunction`.

  Other variants exist:
  - `extern_def?!`: same as `extern_def!?`;
  - `extern_def?`: only defines `myFunction?`;
  - `extern_def!`: only defines `myFunction!`;
  - `extern_def`: defines nothing besides `myFunction`.
-/
scoped syntax (name := externdef)
  declModifiers
  withPosition(
    externKw
    ("in " str)?
    ident declSig (defsItemTail)?
  )
: command

@[inherit_doc externdef, command_elab externdef]
unsafe def externdefImpl : CommandElab
| `(command|
  $mods:declModifiers
  $externKw $[in $path:str]? $ident $sig $[$tail]?
) => do
  let defMod ←
    match externKw with
    | `(externKw| extern_def) => pure none
    | `(externKw| extern_def!) => `(defsMod| !)
    | `(externKw| extern_def?) => `(defsMod| ?)
    | `(externKw| extern_def!?) => `(defsMod| !?)
    | `(externKw| extern_def?!) => `(defsMod| ?!)
    | _ => throwUnsupportedSyntax
  let stx ← `(
    external! $[in $path]?
      $mods:declModifiers
      def $[$defMod]? $ident $sig $[$tail:defsItemTail]?
  )
  Command.elabCommand stx
| _ => throwUnsupportedSyntax

end defsMacro
