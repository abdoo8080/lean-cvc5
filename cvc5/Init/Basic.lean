/-
Copyright (c) 2023-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/



namespace cvc5.Test

/-- Crashes with an explanation if `expected ≠ value`. -/
def check! [Repr α] [BEq α]
  (expected value : α)
: IO Unit :=
  if expected != value then do
    println! "expected `{repr expected}`, got `{repr value}`"
    panic! "check failed"
  else return ()

end cvc5.Test
