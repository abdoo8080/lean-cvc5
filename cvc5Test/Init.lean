import Lean.Server.Utils

import cvc5

namespace cvc5

namespace Test

def IO.run : IO Unit → IO Unit :=
  id

def fail {α : outParam Type} (msg : String) : IO α :=
  IO.throwServerError msg

protected def pref (hint : String) : String :=
  if hint.isEmpty then "" else "[" ++ hint ++ "] "

def assertEq [ToString α] [BEq α] (lft rgt : α) (hint := "") : IO Unit := do
  if lft != rgt then
    IO.eprintln s!"{Test.pref hint}comparison failed: `{lft}` is different from `{rgt}`"
    fail "assertion failed"

def assertTrue (b : Bool) (hint := "") : IO Unit :=
  assertEq true b hint

def assertFalse (b : Bool) (hint := "") : IO Unit :=
  assertEq false b hint

def assertNe [ToString α] [BEq α] (lft rgt : α) (hint := "") : IO Unit := do
  if lft == rgt then
    IO.eprintln s!"{Test.pref hint}comparison failed: `{lft}` is the same as `{rgt}`"
    fail "assertion failed"

def assertOk
  (code : SolverM α)
  (hint : String := "")
: SolverM α := fun solver => do
  match ← code solver with
  | (.ok res, solver) => return (.ok res, solver)
  | (.error e, _) =>
    IO.eprintln s!"{Test.pref hint}expected `.ok` result, got `{e}`"
    fail "assertion failed"

def assertOkDiscard
  (result : SolverM α)
  (hint : String := "")
: SolverM Unit := do
  let _ ← assertOk result hint
  return ()

def assertAnyError
  (expected : String)
  (errorDo : Error → SolverM Unit)
  (result : SolverM α)
  (hint : String := "")
: SolverM Unit := fun solver => do
  match ← result solver with
  | (.ok _, _) =>
    IO.eprintln s!"{Test.pref hint}expected {expected}, got `.ok` result"
    fail "assertion failed"
  | (.error e, solver) => errorDo e solver

def assertError
  (expected : String)
  (result : SolverM α)
  (hint : String := "")
: SolverM Unit :=
  assertAnyError s!"cvc5 error `{expected}`"
    (fun
    | .error err => do
      if err = expected then
        return ()
      else
        IO.eprintln s!"{Test.pref hint}expected cvc5 error `{expected}`, got cvc5 error `{err}`"
        fail "assertion failed"
    | .recoverable err => do
        IO.eprintln s!"{Test.pref hint}expected error `{expected}`, got recoverable error `{err}`"
        fail "assertion failed"
    | .unsupported err => do
        IO.eprintln s!"{Test.pref hint}expected error `{expected}`, got unsupported error `{err}`"
        fail "assertion failed"
    | .option err => do
        IO.eprintln s!"{Test.pref hint}expected error `{expected}`, got option error `{err}`"
        fail "assertion failed"
    | .missingValue => do
        IO.eprintln s!"{Test.pref hint}expected cvc5 error `{expected}`, got missing value error"
        fail "assertion failed"
    )
    result hint

end Test

namespace Result

def toOption (res : Result) : Option Bool :=
  if res.isSat then true
  else if res.isUnsat then false
  else none

end Result

namespace Solver

variable [Monad m]

def checkSat? : SolverT m (Option Bool) :=
  Result.toOption <$> Solver.checkSat

def checkSatAssuming? (terms : Array Term) : SolverT m (Option Bool) :=
  Result.toOption <$> Solver.checkSatAssuming terms

def runWith! [Inhabited α] (tm : TermManager) (query : SolverM α) : IO α := do
  match ← Solver.run tm query with
  | .ok res => return res
  | .error err => IO.throwServerError err.toString

def runIO! [Inhabited α] (query : SolverM α) : IO α := do
  match ← Solver.run (← TermManager.new) query with
  | .ok res => return res
  | .error err => IO.throwServerError err.toString

end Solver

def SolverT.run! [Inhabited α] (query : SolverT IO α) := Solver.runIO! query

namespace Test

scoped syntax
  docComment ?
  "test! " ("[" declId ", " declId "] ")? (ident " => ")? term : command
scoped syntax
  docComment ?
  "test? " ("[" declId ", " declId "] ")? (ident " => ")? term : command

macro_rules
| `(command|
  $[ $outputComment:docComment ]?
  test! $[ [ $fileId:ident , $testId:ident ] ]? $tm:ident => $code:term
) => do
  let errPrefStrLit :=
    match (fileId, testId) with
    | (some fileId, some testId) =>
      Lean.Syntax.mkStrLit s!"[{fileId}.{testId}] test failed:\n"
    | _ => Lean.Syntax.mkStrLit ""
  `(
    $[ $outputComment:docComment ]?
    #guard_msgs in #eval IO.run do
      let $tm:ident ← TermManager.new
      match ← Solver.run $tm (do $code:term) with
      | .ok res => return res
      | .error e => IO.throwServerError ($errPrefStrLit ++ (toString e))
  )
-- | `(command| test! $tm:ident => $code:term) => `(
--   /-- -/
--   #guard_msgs in #eval IO.run do
--     let $tm:ident ← TermManager.new
--     Solver.runWith! $tm do
--       $code:term
-- )
| `(command|
  $[ $outputComment:docComment ]?
  test! $[ [ $fileId:ident , $testId:ident ] ]? $code:term
) => `(
  $[$outputComment]?
  test! $[ [ $fileId, $testId ] ]? _tm => $code
)
| `(command|
  $[ $outputComment:docComment ]?
  test? $[ [ $fileId, $testId ] ]? $tm:ident => $code:term
) => `(
  #eval IO.run do
    let $tm:ident ← TermManager.new
    match ← Solver.run $tm (do $code:term) with
    | .ok res => return res
    | .error e => IO.throwServerError s!"[{fileId}.{testId}] test failed:\n{e}"
)
-- | `(command| $[ $_outputComment:docComment ]? test? $tm:ident => $code:term) => `(
--   #eval IO.run do
--     let $tm:ident ← TermManager.new
--     Solver.runWith! $tm do
--       $code:term
-- )
| `(command| $[$outputComment]? test? $[ [ $fileId, $testId ] ]? $code:term) => `(
  $[$outputComment]?
  test? $[ [ $fileId, $testId ] ]? _tm => $code
)

end Test

end cvc5
