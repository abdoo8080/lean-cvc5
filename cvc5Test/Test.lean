import cvc5



namespace Test

open cvc5 (Solver)

/-- info: bool = `Bool` -/
#guard_msgs in #eval cvc5.runIO do
  let bool ← cvc5.getBooleanSort
  println! "bool = `{bool}`"

/-- error: not a function sort: Bool -/
#guard_msgs in #eval cvc5.runIO do
  let bool ← cvc5.getBooleanSort
  let _sorts ← bool.getFunctionDomainSorts

/-- info: Bool.substitute [Bool → Int] = Int -/
#guard_msgs in #eval cvc5.runIO do
  let bool ← cvc5.getBooleanSort
  let int ← cvc5.getIntegerSort
  let sub ← bool.substitute #[bool] #[int]
  println! "{bool}.substitute [{bool} → {int}] = {sub}"

/-- info: version = `1.2.1` -/
#guard_msgs in #eval cvc5.runIO do
  let solver ← Solver.new
  let version ← solver.getVersion
  println! "version = `{version}`"

/--
info: b1 = `b1`
b2 = `b2`
b1_and_b2 = `(and b1 b2)`
b3 = `b3`
b4 = `b4`
conj = `(and b1 b2 b3 b4)`
-/
#guard_msgs in #eval cvc5.runIO do
  let bool ← cvc5.getBooleanSort
  let solver ← Solver.new
  let b1 ← solver.declareFun "b1" #[] bool
  println! "b1 = `{b1}`"
  let b2 ← solver.declareFun "b2" #[] bool
  println! "b2 = `{b2}`"
  let b1_and_b2 ← cvc5.Term.mk .AND #[b1, b2]
  println! "b1_and_b2 = `{b1_and_b2}`"
  let b3 ← solver.declareFun "b3" #[] bool
  println! "b3 = `{b3}`"
  let b4 ← solver.declareFun "b4" #[] bool
  println! "b4 = `{b4}`"
  let conj ← b1.mkInto .AND #[b2, b3, b4]
  println! "conj = `{conj}`"

/-- error: expecting a Boolean subexpression -/
#guard_msgs in #eval cvc5.runIO do
  let bool ← cvc5.getBooleanSort
  let int ← cvc5.getIntegerSort
  let solver ← Solver.new
  let b1 ← solver.declareFun "b1" #[] bool
  let i1 ← solver.declareFun "i1" #[] int
  let shouldFail ← b1.and i1
  cvc5.throw s!"should have failed: `{shouldFail}`"
  return ()

/-- info:
b1_and_b1 = (and b1 b1)
b1_and_b1_simplified = b1
i1_plus_zero = (+ i1 0)
i1_plus_zero_simplified = i1
-/
#guard_msgs in #eval cvc5.runIO do
  let bool ← cvc5.getBooleanSort
  let int ← cvc5.getIntegerSort
  let solver ← Solver.new

  solver.setOption "produce-models" "true"
  solver.setLogic "QF_LIA"

  let b1 ← solver.declareFun "b1" #[] bool
  let b1_and_b1 ← b1.mkInto .AND #[b1]
  println! "b1_and_b1 = {b1_and_b1}"
  let b1_and_b1_simplified ← solver.simplify b1_and_b1
  println! "b1_and_b1_simplified = {b1_and_b1_simplified}"

  let i1 ← solver.declareFun "i1" #[] int
  let i1_plus_zero ← i1.mkInto .ADD #[← cvc5.Term.mkInteger 0]
  println! "i1_plus_zero = {i1_plus_zero}"
  let i1_plus_zero_simplified ← solver.simplify i1_plus_zero
  println! "i1_plus_zero_simplified = {i1_plus_zero_simplified}"

/-- info:
confirmed unsat
confirmed unsat
confirmed sat
confirmed sat
-/
#guard_msgs in #eval cvc5.runIO do
  let bool ← cvc5.getBooleanSort
  let solver ← Solver.new
  let b1 ← solver.declareFun "b1" #[] bool
  let not_b1 ← b1.mkInto .NOT

  solver.assertFormula b1
  let res ← solver.checkSatAssuming #[not_b1]
  if ¬ res.isUnsat then
    cvc5.throw "expected unsat"
  println! "confirmed unsat"
  solver.assertFormula not_b1
  let res ← solver.checkSat
  if ¬ res.isUnsat then
    cvc5.throw "expected unsat"
  println! "confirmed unsat"

  solver.resetAssertions
  let res ← solver.checkSat
  if ¬ res.isSat then
    cvc5.throw "expected sat"
  println! "confirmed sat"
  let res ← solver.checkSatAssuming #[not_b1]
  if ¬ res.isSat then
    cvc5.throw "expected sat"
  println! "confirmed sat"

  return ()


/-- info:
can't retrieve a symbol created by `parseCommands`
-/
#guard_msgs in #eval cvc5.runIO do
  let solver ← Solver.new
  solver.parseSmtLib "\
(set-logic QF_LIA)
(set-option :produce-models true)

(declare-fun b1 () Bool)
(declare-fun b2 () Bool)

(assert (and b1 b2))
  "
  let res ← solver.checkSat
  if ¬ res.isSat then
    cvc5.throw "expected sat"

  let bool ← cvc5.getBooleanSort
  let b1 ← solver.declareFun "b1" #[] bool (fresh := false)
  solver.assertFormula (← b1.mkInto .NOT)
  let res ← solver.checkSat
  if ¬ res.isSat then cvc5.throw "it works"
  else println! "can't retrieve a symbol created by `parseCommands`"


/-- info:
`(and b1 b2)` sat
unsat after adding `(not b1)`
-/
#guard_msgs in #eval cvc5.runIO do
  let solver ← Solver.new
  let b1_and_b2 := "(and b1 b2)"
  solver.parseSmtLib s!"\
(set-logic QF_LIA)
(set-option :produce-models true)

(declare-fun b1 () Bool)
(declare-fun b2 () Bool)

(assert {b1_and_b2})
  "
  let res ← solver.checkSat
  if ¬ res.isSat
  then cvc5.throw "expected sat"
  else println! "`{b1_and_b2}` sat"

  let not_b1 := "(not b1)"
  solver.parseSmtLib s!"(assert {not_b1})"
  let res ← solver.checkSat
  if ¬ res.isUnsat
  then cvc5.throw "expected unsat"
  else println! "unsat after adding `{not_b1}`"



def findError (s : String) : Option String :=
  let lines := s.trim.splitOn "\n"
  aux lines
where
  aux
    | line :: tail =>
      if line.trimLeft.startsWith "(error" then
        extractError none 0 (line :: tail)
      else
        aux tail
    | [] => none
  extractError (err? : Option String) (paren : Int) : List String → String
    | line :: tail =>
      let paren := parenBalance line paren
      let err := err?.map (s!"{·}\n{line}") |>.getD line
      if paren = 0 then err else extractError err paren tail
    | [] => err? |>.getD "cannot extract parsing error: read EOI"
  parenBalance (s : String) (current : Int) : Int := Id.run do
    let mut balance := current
    for i in [0:s.length] do
      match s.get ⟨i⟩ with
      | '(' => balance := balance + 1
      | ')' => balance := balance - 1
      | _ => pure ()
    return balance


/-- error:
[parsing] Error in option parsing: Argument 'bad' for bool option produce-models is not a bool constant

```output
(error "Error in option parsing: Argument 'bad' for bool option produce-models is not a bool constant")
```
-/
#guard_msgs in #eval cvc5.runIO do
  let solver ← Solver.new
  solver.parseSmtLib s!"\
(set-logic QF_LIA)
(set-option :produce-models bad)
  "

/-- error:
Subexpressions must have the same type:
Equation: (= b i)
Type 1: Bool
Type 2: Int
-/
#guard_msgs in #eval cvc5.runIO do
  let solver ← Solver.new
  solver.parseSmtLib s!"\
(set-logic QF_LIA)
(set-option :produce-models true)

(declare-fun b () Bool)
(declare-fun i () Int)

(assert (= b i))
  "

/-- error:
[parsing] cannot get model unless after a SAT or UNKNOWN response.

```output
unsat
(error "cannot get model unless after a SAT or UNKNOWN response.")
```
-/
#guard_msgs in #eval cvc5.runIO do
  let solver ← Solver.new
  solver.parseSmtLib s!"\
(set-logic QF_LIA)
(set-option :produce-models true)

(declare-fun b () Bool)

(assert (and b (not b)))

(check-sat)
(get-model)
  "
