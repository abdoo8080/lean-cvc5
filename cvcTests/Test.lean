import cvc5



namespace Test

/-- info: bool = `Bool` -/
#guard_msgs in #eval Cvc5.runIO do
  let bool ← Cvc5.getBooleanSort
  println! "bool = `{bool}`"

/-- error: [error] not a function sort: Bool -/
#guard_msgs in #eval Cvc5.runIO do
  let bool ← Cvc5.getBooleanSort
  let _sorts ← bool.getFunctionDomainSorts

/-- info: Bool.substitute [Bool → Int] = Int -/
#guard_msgs in #eval Cvc5.runIO do
  let bool ← Cvc5.getBooleanSort
  let int ← Cvc5.getIntegerSort
  let sub ← bool.substitute #[bool] #[int]
  println! "{bool}.substitute [{bool} → {int}] = {sub}"

/-- info: x = `x` -/
#guard_msgs in #eval Cvc5.runIO do
  let bool ← Cvc5.getBooleanSort
  let solver ← Cvc5.Solver.new
  let x ← solver.declareFun "x" #[] bool
  println! "x = `{x}`"
