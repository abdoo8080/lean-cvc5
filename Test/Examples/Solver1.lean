import Test.Init

namespace cvc5.Test

def work : IO Unit := Smt.run! do
  Smt.setOption "produce-proofs" "true"

  Smt.setLogic' Logic.qf_lia

  let n1 ← Smt.declareConst "n1" Int
  let n2 ← Smt.declareConst "n2" Int

  let b ← Smt.declareConst "b" Bool

  let eq ← Smt.mkEqual n1 n2
  let neq ← Smt.mkNot eq
  let ite ← Smt.mkIte b eq neq

  Smt.assertFormula ite
  Smt.assertFormula eq

  match ← Smt.checkSat? with
  | none =>
    panic! "got a timeout"
  | some false =>
    panic! "unexpected `unsat` result"
  | some true =>
    println! "confirmed `sat` result"

  Smt.resetAssertions

  let imp1 ← Smt.mkImplies b eq
  let not_b ← Smt.mkNot b
  let imp2 ← Smt.mkImplies not_b neq

  Smt.assertFormula imp1
  Smt.assertFormula imp2
  Smt.assertFormula eq
  Smt.assertFormula not_b

  match ← Smt.checkSat? with
  | none => panic! "got a timeout"
  | some false => println! "confirmed `unsat` result"
  | some true => panic! "unexpected `sat` result"

  let proofs ← Smt.getProof

  println! "proof:"
  for p in proofs do
    println! "- {p.getResult}"

#eval work
