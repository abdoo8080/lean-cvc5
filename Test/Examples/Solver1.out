confirmed `sat` result
confirmed `unsat` result
proof:
- (let ((_let_1 (not b))) (let ((_let_2 (= n1 n2))) (not (and (=> b _let_2) (=> _let_1 (not _let_2)) _let_2 _let_1))))